pinot-core/src/main/java/org/apache/pinot/core/util/GenericRowSerDeUtils.java

```diff
@@ -0,0 +1,326 @@
+          case LONG:
+            for (Object element : multiValue) {
+              byteBuffer.putLong((long) element);
+            }
+            break;
+          case FLOAT:
+            for (Object element : multiValue) {
+              byteBuffer.putFloat((float) element);
+            }
+            break;
+          case DOUBLE:
+            for (Object element : multiValue) {
+              byteBuffer.putDouble((double) element);
+            }
+            break;
+          case BOOLEAN:
+          case STRING:
+            for (Object element : multiValue) {
+              byte[] stringBytes = ((String) element).getBytes(StandardCharsets.UTF_8);
+              byteBuffer.putInt(stringBytes.length);
+              byteBuffer.put(stringBytes);
+            }
+            break;
+          case BYTES:
+            for (Object element : multiValue) {
+              byte[] bytes = (byte[]) element;
+              byteBuffer.putInt(bytes.length);
+              byteBuffer.put(bytes);
+            }
+            break;
+          default:
+            throw new UnsupportedOperationException(
+                String.format("DataType '%s' not supported", fieldSpec.getDataType()));
+        }
+      }
+    }
+    return genericRowBytes;
+  }
+
+  /**
+   * Deserializes bytes from the buffer to GenericRow
+   * @param dataBuffer the pinot data buffer
+   * @param offset offset to begin reading from
+   * @param fieldSpecs list of field specs to determine fields in deserialization
+   * @param reuse GenericRow object for returning
+   * @return Deserialized GenericRow
+   */
+  public static GenericRow deserializeGenericRow(PinotDataBuffer dataBuffer, long offset, List<FieldSpec> fieldSpecs,
+      GenericRow reuse) {
+    for (FieldSpec fieldSpec : fieldSpecs) {
+
+      if (fieldSpec.isSingleValueField()) {
+        switch (fieldSpec.getDataType().getStoredType()) {
+
+          case INT:
+            int intValue = dataBuffer.getInt(offset);
+            reuse.putValue(fieldSpec.getName(), intValue);
+            offset += Integer.BYTES;
+            break;
+          case LONG:
+            long longValue = dataBuffer.getLong(offset);
+            reuse.putValue(fieldSpec.getName(), longValue);
+            offset += Long.BYTES;
+            break;
+          case FLOAT:
+            float floatValue = dataBuffer.getFloat(offset);
+            reuse.putValue(fieldSpec.getName(), floatValue);
+            offset += Float.BYTES;
+            break;
+          case DOUBLE:
+            double doubleValue = dataBuffer.getDouble(offset);
+            reuse.putValue(fieldSpec.getName(), doubleValue);
+            offset += Double.BYTES;
+            break;
+          case STRING:
+            int stringSize = dataBuffer.getInt(offset);
+            offset += Integer.BYTES;
+            byte[] stringBytes = new byte[stringSize];
+            for (int j = 0; j < stringSize; j++) {
+              stringBytes[j] = dataBuffer.getByte(offset);
```

Comment:
Use `copyTo()` to read multiple bytes. Same for BYTES type and MV fields

Reply:
ah nice hadn't noticed that API. Changed it for string and bytes in both SV and MV cases.
Did you mean MV for all other data types too? I still need to fetch each individual element in the array as `getInt` `getFloat` etc

Reply:
Just STRING and BYTES